tools/PI/DevHome.PI/Helpers/SnapHelper.cs

```diff
@@ -0,0 +1,149 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.ComponentModel;
+using System.Diagnostics;
+using CommunityToolkit.Mvvm.ComponentModel;
+using DevHome.Common.Extensions;
+using DevHome.PI.Models;
+using DevHome.PI.ViewModels;
+using Microsoft.UI.Xaml;
+using Windows.Graphics;
+using Windows.Win32;
+using Windows.Win32.Foundation;
+using Windows.Win32.UI.Accessibility;
+using WinUIEx;
+
+namespace DevHome.PI.Helpers;
+
+public class SnapHelper
+{
+    // TODO The SnapOffsetHorizontal and UnsnapGap values don't allow for different DPIs.
+    private const int UnsnapGap = 9;
+
+    // It seems the way rounded corners are implemented means that the window is really 8px
+    // bigger than it seems, so we'll subtract this when we do sidecar snapping.
+    private const int SnapOffsetHorizontal = 8;
+
+    private readonly WINEVENTPROC _winPositionEventDelegate;
+    private readonly WINEVENTPROC _winFocusEventDelegate;
+
+    private HWINEVENTHOOK _positionEventHook;
+    private HWINEVENTHOOK _focusEventHook;
+
+    public SnapHelper()
+    {
+        _winPositionEventDelegate = new(WinPositionEventProc);
+        _winFocusEventDelegate = new(WinFocusEventProc);
+    }
+
+    public void Snap()
+    {
+        Debug.Assert(_positionEventHook == HWINEVENTHOOK.Null, "Hook should be cleared");
+        Debug.Assert(_focusEventHook == HWINEVENTHOOK.Null, "Hook should be cleared");
+
+        _positionEventHook = WindowHelper.WatchWindowPositionEvents(_winPositionEventDelegate, (uint)TargetAppData.Instance.ProcessId);
+        _focusEventHook = WindowHelper.WatchWindowFocusEvents(_winFocusEventDelegate, (uint)TargetAppData.Instance.ProcessId);
+
+        SnapToWindow();
+    }
+
+    public void Unsnap()
+    {
+        var barWindow = Application.Current.GetService<PrimaryWindow>().DBarWindow;
```

Comment:
Can this class be a Service? Then you can use dependency injection to get it into BarWindowViewModel, and use DI to get the PrimaryWindow service throughout here.

Reply:
Yes, the goal is to get Barwindow to be a service. We have some more cleanup to be done before I can do that and this PR is a part of the cleanup.